### DISCUSSION TITLE
The examples in the package do not build

### DISCUSSION THREAD

**Discussion by ghost** [hathach/tinyusb#1886] [Q&A]
Hi,
I am trying to build the examples inside the package following step by step your procedure.
I use a Mac M1 running Ventura.
In a short:

git clone https://github.com/hathach/tinyusb tinyusb
cd tinyusb
git submodule update --init lib
cd examples/device/cdc_msc

make BOARD=stm32l4r5nucleo get-deps
make BOARD=stm32l4r5nucleo all

So, it compiles some parts of the project but then, it stops at the link

CC video_device.o
CC vendor_device.o
LINK _build/stm32l4r5nucleo/cdc_msc.elf
/opt/uKOS/cross/gcc-11.3.0/cortex-M/bin/../lib/gcc/arm-none-eabi/11.3.1/../../../../arm-none-eabi/bin/ld: cannot find _build/stm32l4r5nucleo/obj//src/main.o: No such file or directory
/opt/uKOS/cross/gcc-11.3.0/cortex-M/bin/../lib/gcc/arm-none-eabi/11.3.1/../../../../arm-none-eabi/bin/ld: cannot find _build/stm32l4r5nucleo/obj//src/msc_disk.o: No such file or directory
/opt/uKOS/cross/gcc-11.3.0/cortex-M/bin/../lib/gcc/arm-none-eabi/11.3.1/../../../../arm-none-eabi/bin/ld: cannot find _build/stm32l4r5nucleo/obj//src/usb_descriptors.o: No such file or directory
collect2: error: ld returned 1 exit status
make: *** [_build/stm32l4r5nucleo/cdc_msc.elf] Error 1

For sure I have to miss some evident steps, but I cannot figure out the ones.
This happens with all the included projects.
Any suggestion to fix it?
BR

**Comment by hathach** [comment]
this is weird, it should be good. I am not sure what is wrong either, unfortunately I am not using macos and couldn't look further at this. Hopefully other users would help.

**Reply by ghost** [reply]
Hi, I solved the problem by including this line in the Project Makefil

CURRENT_PATH=examples/device/cdc_msc

(for example for the cdc_msc example)
BR
Edo

**Comment by aspdigital** [comment]
Just to chime in with support, I'm getting a similar error with an lpcxpresso55s28 board. I am working with the midi_test example. I am using an M1 Pro Mac running Ventura, but I don't think it's a problem with the computer or the OS, as such. My path points to the ARM gcc as installed with the MCUxpresso IDE.

The make output is:

```
andy@mitzvah midi_test % make BOARD=lpcxpresso55s28 all     
realpath: illegal option -- -
usage: realpath [-q] [path ...]
"PORT1 High Speed"
AS startup_LPC55S28_asm.o
CC dcd_lpc_ip3511.o
CC system_LPC55S28.o
CC fsl_clock.o
/Users/andy/Projects/tinyusb/hw/mcu/nxp/mcux-sdk/devices/LPC55S28/drivers/fsl_clock.c: In function 'CLOCK_EnableUsbhs0PhyPllClock':
/Users/andy/Projects/tinyusb/hw/mcu/nxp/mcux-sdk/devices/LPC55S28/drivers/fsl_clock.c:1986:56: warning: unused parameter 'src' [-Wunused-parameter]
 1986 | bool CLOCK_EnableUsbhs0PhyPllClock(clock_usb_phy_src_t src, uint32_t freq)
      |                                    ~~~~~~~~~~~~~~~~~~~~^~~
/Users/andy/Projects/tinyusb/hw/mcu/nxp/mcux-sdk/devices/LPC55S28/drivers/fsl_clock.c: In function 'CLOCK_EnableUsbhs0DeviceClock':
/Users/andy/Projects/tinyusb/hw/mcu/nxp/mcux-sdk/devices/LPC55S28/drivers/fsl_clock.c:2074:54: warning: unused parameter 'src' [-Wunused-parameter]
 2074 | bool CLOCK_EnableUsbhs0DeviceClock(clock_usbhs_src_t src, uint32_t freq)
      |                                    ~~~~~~~~~~~~~~~~~~^~~
/Users/andy/Projects/tinyusb/hw/mcu/nxp/mcux-sdk/devices/LPC55S28/drivers/fsl_clock.c:2074:68: warning: unused parameter 'freq' [-Wunused-parameter]
 2074 | bool CLOCK_EnableUsbhs0DeviceClock(clock_usbhs_src_t src, uint32_t freq)
      |                                                           ~~~~~~~~~^~~~
/Users/andy/Projects/tinyusb/hw/mcu/nxp/mcux-sdk/devices/LPC55S28/drivers/fsl_clock.c: In function 'CLOCK_EnableUsbhs0HostCloc':
/Users/andy/Projects/tinyusb/hw/mcu/nxp/mcux-sdk/devices/LPC55S28/drivers/fsl_clock.c:2086:52: warning: unused parameter 'src' [-Wunused-parameter]
 2086 | bool CLOCK_EnableUsbhs0HostClock(clock_usbhs_src_t src, uint32_t freq)
      |                                  ~~~~~~~~~~~~~~~~~~^~~
/Users/andy/Projects/tinyusb/hw/mcu/nxp/mcux-sdk/devices/LPC55S28/drivers/fsl_clock.c:2086:66: warning: unused parameter 'freq' [-Wunused-parameter]
 2086 | bool CLOCK_EnableUsbhs0HostClock(clock_usbhs_src_t src, uint32_t freq)
      |                                                         ~~~~~~~~~^~~~
CC fsl_power.o
CC fsl_reset.o
CC fsl_gpio.o
/Users/andy/Projects/tinyusb/hw/mcu/nxp/mcux-sdk/drivers/lpc_gpio/fsl_gpio.c: In function 'GPIO_EnablePortClock':
/Users/andy/Projects/tinyusb/hw/mcu/nxp/mcux-sdk/drivers/lpc_gpio/fsl_gpio.c:42:45: warning: unused parameter 'base' [-Wunused-parameter]
   42 | static void GPIO_EnablePortClock(GPIO_Type *base, uint32_t port)
      |                                  ~~~~~~~~~~~^~~~
CC fsl_common_arm.o
CC fsl_flexcomm.o
CC fsl_usart.o
/Users/andy/Projects/tinyusb/hw/mcu/nxp/mcux-sdk/drivers/flexcomm/fsl_usart.c: In function 'USART_TransferGetReceiveCount':
/Users/andy/Projects/tinyusb/hw/mcu/nxp/mcux-sdk/drivers/flexcomm/fsl_usart.c:992:52: warning: unused parameter 'base' [-Wunused-parameter]
  992 | status_t USART_TransferGetReceiveCount(USART_Type *base, usart_handle_t *handle, uint32_t *count)
      |                                        ~~~~~~~~~~~~^~~~
CC sct_neopixel.o
/Users/andy/Projects/tinyusb/lib/sct_neopixel/sct_neopixel.c: In function 'sctpix_init':
/Users/andy/Projects/tinyusb/lib/sct_neopixel/sct_neopixel.c:150:27: warning: unused parameter 'neoPixelType' [-Wunused-parameter]
  150 | void sctpix_init(uint32_t neoPixelType) {
      |                  ~~~~~~~~~^~~~~~~~~~~~
CC family.o
CC board.o
CC tusb.o
CC tusb_fifo.o
CC usbd.o
CC usbd_control.o
CC audio_device.o
CC cdc_device.o
CC dfu_device.o
CC dfu_rt_device.o
CC hid_device.o
CC midi_device.o
CC msc_device.o
CC ecm_rndis_device.o
CC ncm_device.o
CC usbtmc_device.o
CC video_device.o
CC vendor_device.o
LINK _build/lpcxpresso55s28/midi_test.elf
arm-none-eabi-gcc: error: _build/lpcxpresso55s28/obj//src/main.o: No such file or directory
arm-none-eabi-gcc: error: _build/lpcxpresso55s28/obj//src/usb_descriptors.o: No such file or directory
make: *** [_build/lpcxpresso55s28/midi_test.elf] Error 1
```
And the linker fail tells me that it's not compiling either main.c or usb_descriptors.c, which you can see from the make result. And yeah, look through the compiles and neither are mention. What's odd is that the makefile seems to think that main.c and usb_descriptors.c live in the source directory for the board, not for the test example which is where I started the build.

I'm getting lost in all of the makefile inclusion. Seems like the thing _should_ build the two sources in the example. I don't know why it fails.

**Reply by aspdigital** [reply]
I added the helpful $(info ) to the example's Makefile:

`SRC_C +=  $(addprefix $(CURRENT_PATH)/, $(EXAMPLE_SOURCE))`

`$(info SRC_C is $(SRC_C))`

and this gives:

`SRC_C is src/portable/nxp/lpc_ip3511/dcd_lpc_ip3511.c hw/mcu/nxp/mcux-sdk/devices/LPC55S28/system_LPC55S28.c hw/mcu/nxp/mcux-sdk/devices/LPC55S28/drivers/fsl_clock.c hw/mcu/nxp/mcux-sdk/devices/LPC55S28/drivers/fsl_power.c hw/mcu/nxp/mcux-sdk/devices/LPC55S28/drivers/fsl_reset.c hw/mcu/nxp/mcux-sdk/drivers/lpc_gpio/fsl_gpio.c hw/mcu/nxp/mcux-sdk/drivers/common/fsl_common_arm.c hw/mcu/nxp/mcux-sdk/drivers/flexcomm/fsl_flexcomm.c hw/mcu/nxp/mcux-sdk/drivers/flexcomm/fsl_usart.c lib/sct_neopixel/sct_neopixel.c hw/bsp/lpc55/family.c hw/bsp/board.c  /src/main.c /src/usb_descriptors.c`

and we see at the end that the two sources for the example get their path set to the root and they're never compiled.

Eliminating the addprefix and instead updating the source list as

`SRC_C += $(EXAMPLE_SOURCE)`

sets the path to those sources correctly, so they build and the project links.

If I was clever enough to know how to submit a PR I would do so.

Edited to add: I was able to load the build result into my LPC55S28 board with a J-Link and the program works: Pocket MIDI shows the example's periodic note-on/note-off sequences.

The machine is a 2021 16" MacBook Pro with the M1 Pro CPU running the latest Ventura 13.2.1.

**Comment by hathach** [comment]
should be fixed by https://github.com/hathach/tinyusb/pull/1938 , will merge shortly, check it out to see if that works for you
